mitmdump: temporarily running out of diskspace corrupts dumpfile

sw9719 opened the pull request:
#### Description

As mentioned in this [issue](https://github.com/mitmproxy/mitmproxy/issues/4460), mitmdump writing to a file can lead to data corruption when disk becomes full.
To resolve this, this PR adds disk usage checks and exits the program when it exceeds a preset threshold. 

#### Checklist

 - [ ] If this is acceptable, I can write tests for this.

sw9719 commented:
Hello, just wanted to check if this is acceptable

mhils commented on mitmproxy/io/tnetstring.py:
```diff
@@ -63,7 +66,14 @@ def dump(value: TSerializable, file_handle: typing.IO[bytes]) -> None:
+            ctx.log.error("Exiting due to insufficient space on disk")
+            sys.exit(1)
+        else:
+            raise e
```
The logic here looks good, but this should really only happen when continuously streaming flows to disk. 

Imagine you have an interactive mitmproxy session, you want to save your flows, there is not enough disk space, and now everything is closed automatically. That would be painful. :) Long story short, let's move this into https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/addons/save.py.

sw9719 replied:
When running mitmproxy,  if a user decides to stream to a file, it will still exit without a warning because the function executed will be same for both mitmproxy and mitmdump. I think I need to look for a way to detect if the 'save stream to file' was requested by mitmproxy or mitmdump.

mhils replied:
Yes, the save addon handles that for both. :)

sw9719 commented:
Since `stream.add` is called from multiple methods in `addons/save.py`, I added the excpetion handling in `io/io.py`. Also added logic to detect if the `Save` addon was initialized for mitmdump or mitmproxy and tested it successfully.

mhils reviewed (changes requested):
`self.stream.add` is called four times in the save addon. I think that's a reasonable amount to cover, feel free to write a helper function there if you want to reuse functionality. The current approach here introduces unnecessary complexity across multiple interfaces.

sw9719 commented:
I am a bit confused. Do you mean create a wrapper function for `stream.add` and add the mitmdump detection logic in that?

mhils commented:
We don't want to detect the usage of mitmdump, we specifically want to detect the usage of `save_stream_file` (which sets `self.stream` in the Save addon).

mhils commented on mitmproxy/addons/save.py:
```diff
@@ -44,6 +46,17 @@ def start_stream_to_path(self, path, flt):
+        except OSError as e:
+            if e.errno == errno.ENOSPC:
+                ctx.log.error("Exiting due to insufficient space on disk")
+                if isinstance(self.stream, io.FilteredFlowWriter):
```
I think we can also remove this check? Is there a case where this method is called and `self.stream` is not a FilteredFlowWriter?

sw9719 replied:
Thinking about it now, it is not needed. This check is from one of my previous commits as it was breaking on of the tests. Ill remove it and re-run the tests to make sure nothing breaks.

mhils commented on mitmproxy/addons/save.py:
```diff
@@ -44,6 +46,17 @@ def start_stream_to_path(self, path, flt):
+        try:
+            self.stream.add(flow)
+        except OSError as e:
+            if e.errno == errno.ENOSPC:
```
Practically speaking it may be useful if we just exit on _every_ OSError? I mean in any case it indicates that we cannot save anymore, which is critical.

sw9719 replied:
Okay, but if exit on every OSError, then we might not understand why exactly the call failed. Is it better to add log the error string from OSError and then exit?

mhils replied:
Yes, that sounds great.

sw9719 replied:
Hello, I fixed the code as needed. Is this ok? If yes, Ill proceed with tests.

mhils reviewed (commented):
This looks great now, thanks for incorporating the feedback! :cake:

I have two smaller remarks, other than that we need a test and we should be good! :)

sw9719 commented:
For the test, I was thinking of creating a small file system and mounting it on test directory and filing it. But this require root privileges so is that fine or do we need some other method?

mhils commented:
I think this is addressed by https://github.com/mitmproxy/mitmproxy/commit/3a5550a09cd40d76acfe71aa45c7a8309525ad51 now. Thanks again! 🍰
